src_c/rect.c

```diff
@@ -572,21 +572,41 @@ pg_rect_unionall_ip(pgRectObject *self, PyObject *args)
 }
 
 static PyObject *
-pg_rect_collidepoint(pgRectObject *self, PyObject *args)
+pg_rect_collidepoint(pgRectObject *self, PyObject *const *args,
+                     Py_ssize_t nargs)
 {
-    int x = 0, y = 0;
-    int inside;
+    SDL_Rect srect = self->r;
+    SDL_Point p;
 
-    if (!pg_TwoIntsFromObj(args, &x, &y)) {
-        return RAISE(PyExc_TypeError, "argument must contain two numbers");
+    if (nargs == 1) {
+        if (!pg_TwoIntsFromObj(args[0], &(p.x), &(p.y)))
+            return RAISE(PyExc_TypeError, "argument must contain two numbers");
     }
+    else if (nargs == 2) {
+        p.x = PyLong_AsLong(args[0]);
+        if (PyErr_Occurred()) {
+            PyErr_Clear();
+            p.x = (int)PyFloat_AsDouble(args[0]);
+            if (PyErr_Occurred())
+                return NULL;
+        }
```

Comment:
I'd suggest using `pg_IntFromObj` here (it is what `pg_TwoIntsFromObj` uses internally)

Reply:
I know but this is 12.5% faster than using pg_IntFromObj, and i believe that we should squeeze all perf from a function like this.

Reply:
Also i'll put out more tests and a new commit in about 8 hours, With much improved performance and a comparison with and without `pg_IntFromObj`. in some cases using `pg_IntFromObj` (in my new commit) reduced performance by 30%. I'll probably make a fast `pg_FastIntFromObj` if you want clean code, but for now `pg_IntFromObj` is simply not the right choice.

Reply:
The problem I see with the current code is that if `PyLong_AsLong` fails with an error that is not because it got a float, that error will be cleared (instead of being forwarded to the user). I also don't like adding code repetition, it will definitely add a maintenance burden in the long term. Adding micro-optimisations to centralised util functions is a better alternative because that will benefit many more functions using it.

Reply:
>  if `PyLong_AsLong` fails with an error that is not because it got a float, that error will be cleared (instead of being forwarded to the user)

This is already happening in `pg_IntFromObj`, just the other way around, it first checks for float, if so converts to double (dunno why not  `PyFloat_AS_DOUBLE`) and casts to int, else tries to convert to long with no check. If it errors, the error is cleared and the user has to invent an error. In this case `pg_IntFromObj` has worse error checking than what i have as it literally discards precise error messages in any case, while i at least forward one precise error message in one case.

```C
static int
pg_IntFromObj(PyObject *obj, int *val)
{
    int tmp_val;

    if (PyFloat_Check(obj)) {
        /* Python3.8 complains with deprecation warnings if we pass
         * floats to PyLong_AsLong.
         */
        double dv = PyFloat_AsDouble(obj);
        tmp_val = (int)dv;
    }
    else {
        tmp_val = PyLong_AsLong(obj);
    }

    if (tmp_val == -1 && PyErr_Occurred()) {
        PyErr_Clear();
        return 0;
    }
    *val = tmp_val;
    return 1;
}
```